### PR TITLE
Make it easier to add options for custom executables

### PR DESCRIPTION
**Pull request by ugermann**
So far, addition options to a marian-based executable required changes to ConfigParser. With respect to custom marian-based executables (websocket server, REST server, RabbitMQ worker, Kafka Worker, etc.), this design is bound to lead to problems (just lock at the cli::mode::server hack).

This pull request includes  a backward-compatible refactoring of the way options are handled. With these changes, we can, for example, do the following:

```C++
int main(int argc, char* argv[]) {
  ConfigParser cp(cli::mode::translation);
  cp.addOption<int>("--port", // option name
                     "Server Options", // option group name
                     "Port for server.", // help string
                     5678); // default value
  auto opts = cp.parseOptions(argc,argv,/*validate=*/true);
  ...
```

Major Changes:
- CliWrapper::switchGroup() now returns the value of the current group before the switch. This allows us to switch back to the previous group after adding options to a group.
- new ConfigParser constructor that adds core options at construction; mostly a code move from the beginning of ConfigParser::parseOptions(). 
- ConfigParser::parseOptions() returns a Ptr\<Options> with the options as parsed.

### CODE REVIEW

**Comment by snukky** [comment]
That's a nice improvement. With regard to the third change, if I remember correctly, one of the reasons why the Options class has been introduced (already having a very similar Config class) was making it independent from Config, ConfigParser, etc. as these were planned to be rewritten (e.g. #305). Maybe we should try to avoid having a ConfigParser in Options?

**Comment by ugermann** [comment]
I've changed things around so that ConfigParser::parseOptions() now returns a new Options instance with the options as parsed. The options class is now independent of Config and ConfigParser.

**Comment by ugermann** [comment]
I think it would be really worthwhile to merge this pull request before the MT Marathon starts and before everyone starts adding their custom options to the ConfigParser class.

**Comment by emjotde** [comment]
ping @snukky for review.

File: src/common/config_parser.h [hunk 1]
```diff
@@ -25,12 +25,49 @@ enum struct mode { training, translation, scoring, server };
...
+    cli_.switchGroup(previous_group);
+    return *this;
+  }
+
+  template<typename T>
+  ConfigParser&
+  addOption(const std::string& args,
+            const std::string& group,
+            const std::string& help,
+            const T val, const T implicit_val) {
```
Comment (snukky) [inline src/common/config_parser.h]: The code is organized vertically, but there are two arguments in one line. Lines 49-50 can be merged to safe the vertical space instead ;)

File: src/common/config_parser.h [hunk 2]
```diff
@@ -25,12 +25,49 @@ enum struct mode { training, translation, scoring, server };
...
+            const T val) {
+    std::string previous_group = cli_.switchGroup(group);
+    cli_.add<T>(args,help,val);
+    cli_.switchGroup(previous_group);
+    return *this;
+  }
+
+  template<typename T>
+  ConfigParser&
+  addOption(const std::string& args,
```
Comment (snukky) [inline src/common/config_parser.h]: Maybe both `addOption` methods can be replaced with a method returning reference to `cli_` and let the user to call any of `.add<T>` variations?

File: src/common/config_parser.h [hunk 3]
```diff
@@ -47,15 +84,19 @@ class ConfigParser {
...
-
-  YAML::Node getConfig() const;
 
+  Ptr<Options> parseOptions(int argc, char** argv, bool validate);
+  YAML::Node const& getConfig() const;
+  cli::mode getMode() const;
+  std::string const& cmdLine() const;
 private:
-  bool modeServer_;
+  // bool modeServer_;
```
Comment (snukky) [inline src/common/config_parser.h]: Can we remove this?

File: src/common/config_parser.cpp [hunk 4]
```diff
@@ -48,6 +50,57 @@ const std::set<std::string> PATHS = {
...
+}
+
+std::string const& ConfigParser::cmdLine() const {
+  return cmdLine_;
+}
+
+ConfigParser::ConfigParser(cli::mode mode)
+  : cli_(config_,"Marian: Fast Neural Machine Translation in C++",
+         "General options", "", 40),
+    // modeServer_(mode == cli::mode::server),
```
Comment (snukky) [inline src/common/config_parser.cpp]: Can we remove the comment?

File: src/common/config.h [hunk 5]
```diff
@@ -47,7 +48,8 @@ class Config {
   Config(const Config& other);
   Config(const Options& options);
 
-  void initialize(int argc, char** argv, cli::mode mode, bool validate);
+  void initialize(ConfigParser const& cp);
+  // void initialize(int argc, char** argv, cli::mode mode, bool validate);
```
Comment (snukky) [inline src/common/config.h]: Commented code

File: src/common/config_parser.h [hunk 6]
```diff
@@ -25,12 +25,49 @@ enum struct mode { training, translation, scoring, server };
  */
 class ConfigParser {
 public:
+
+  ConfigParser(cli::mode mode);
```
Comment (snukky) [inline src/common/config_parser.h]: Could you add a short usage example of this constructor, similar to one in your description of this PR?

File: src/common/options.h [hunk 7]
```diff
@@ -32,22 +32,20 @@ class Options {
   YAML::Node options_;
 
 public:
-  Options() {}
-  Options(const Options& other) : options_(YAML::Clone(other.options_)) {}
+  Options();
+  Options(const Options& other);
+  // Options(ConfigParser& cp, int argc, char** argv, bool validate);
```
[the remark this answers is not in the document]
Comment (ugermann) [reply src/common/options.h]: Done.

File: src/common/options.cpp [hunk 8]
```diff
@@ -0,0 +1,62 @@
+// -*- mode: c++; indent-tabs-mode: nil; tab-width: 2 -*-
```
[the remark this answers is not in the document]
Comment (ugermann) [reply src/common/options.cpp]: Done.

File: src/common/config_parser.h [hunk 9]
```diff
@@ -1,3 +1,4 @@
+// -*- mode: c++; indent-tabs-mode: nil; tab-width: 2 -*-
```
[the remark this answers is not in the document]
Comment (ugermann) [reply src/common/config_parser.h]: Done.

File: src/common/config.cpp [hunk 10]
```diff
@@ -1,4 +1,6 @@
 #include "common/config.h"
+#include "common/options.h"
+#include "common/config_parser.h"
```
[the remark this answers is not in the document]
Comment (ugermann) [reply src/common/config.cpp]: Done.

File: src/common/config_parser.cpp [hunk 11]
```diff
@@ -5,8 +5,10 @@
 #include "common/config_validator.h"
 #include "common/file_stream.h"
 #include "common/logging.h"
+#include "common/options.h"
+#include "common/config.h"
 #include "common/utils.h"
-
+#include "common/regex.h"
```
[the remark this answers is not in the document]
Comment (ugermann) [reply src/common/config_parser.cpp]: Done.

File: src/common/config_parser.h (hunk 2 above)
Comment (ugermann) [reply src/common/config_parser.h]: That's an invitation for developers to break things. See my comments on cli::switchGroup below.

File: src/common/config_parser.h (hunk 3 above)
Comment (ugermann) [reply src/common/config_parser.h]: Yes. Done.

File: src/common/config_parser.cpp (hunk 4 above)
Comment (ugermann) [reply src/common/config_parser.cpp]: Done.

File: src/common/config.h (hunk 5 above)
Comment (ugermann) [reply src/common/config.h]: Done.

File: src/common/config_parser.h (hunk 6 above)
Comment (ugermann) [reply src/common/config_parser.h]: Done in comments above class definition.

**Comment by ugermann** [comment]
Merged with latest version of master branch.
Results of regression tests (on frigg):
```
Skipped:
  - tests/examples/iris/test_iris.sh
  - tests/examples/mnist/test_mnist_ffnn.sh
  - tests/training/features/exp-smoothing/test_expsmooth_sync.sh
  - tests/training/multi-gpu/test_async_sgd_runs.sh
  - tests/training/multi-gpu/test_sync_sgd.sh
  - tests/training/restoring/optimizer/test_adam_params_async.sh
  - tests/training/restoring/optimizer/test_adam_params_sync.sh
  - tests/training/restoring/multi-gpu/test_async.sh
  - tests/training/restoring/multi-gpu/test_sync.sh
  - tests/training/restoring/exp-smoothing/test_expsmooth_sync.sh
Failed:
  - tests/training/cost-functions/test_perplexity.sh
Logs:
  - /raid0/germann/regtests/marian.ug-command-line-options2/regression-tests/tests/training/cost-functions/test_perplexity.sh.log
```

Tested on frigg. I have no idea why the perplexity test fails.

**Comment by emjotde** [comment]
That one is oddly instable.

**Comment by ugermann** [comment]
pinging this pull request

**Pull request merged**